File: 001_Analoge_Schaltungen/KiCad_Biquad/3_order_lp/redpitaya_scpi/bode_data_meas.m
%% STEMlab measurements for Bode plots
% @author: Mirco Meiners (HSB)
% Input signal: DF_IN1
% Output signal: DF_IN2
clear;

%% Define Red Pitaya as TCP client object
IP = '192.168.111.183';  % IP of your Red Pitaya ...
port = 5000;
RP = tcpclient(IP, port);
RP.InputBufferSize = 16384*32;

%% Open connection to Red Pitaya
RP.ByteOrder = "big-endian";
configureTerminator(RP, "CR/LF");

flush(RP);

%% Generate continous signal
func = "SINE";  % {sine, square, triangle, sawu, sawd, pwm}
ampl = 0.5;  % Set amplitude
offset = 0.0;  % Set offset
freqs = [500:100:2000];  % Set frequencies


%% Loop to measure multiple tones
for n = 1:length(freqs)
    % Send SCPI command to Red Pitaya to turn ON generator
    writeline(RP,'GEN:RST');  % Reset Generator
    writeline(RP,strcat("SOUR1:FUNC ", func));  % Set function of output signal
    writeline(RP,strcat("SOUR1:VOLT ", num2str(ampl)));  % Set amplitude
    writeline(RP,strcat("SOUR1:VOLT:OFFS ", num2str(offset)));  % Set offset
    writeline(RP,strcat("SOUR1:FREQ:FIX ", num2str(freqs(n))));  % Set frequency
    writeline(RP,'OUTPUT1:STATE ON');  % Turn on output OUT2
    
    writeline(RP,'SOUR1:TRig:INT');  % Generate trigger
    
    pause(1);

    % Trigger
    writeline(RP,'ACQ:RST');  % Input reset
    writeline(RP,'ACQ:DATA:FORMAT ASCII')
    writeline(RP,'ACQ:DATA:UNITS VOLTS')
    writeline(RP,'ACQ:DEC 64');  % Decimation 64
    writeline(RP,'ACQ:TRIG:LEV 0.5');  % Trigger level
    
    % Set trigger delay to 0 samples
    % 0 samples delay sets trigger to the center of the buffer
    % Signal on your graph will have the trigger in the center (symmetrical)
    % Samples from left to the center are samples before trigger
    % Samples from center to the right are samples after trigger
    
    writeline(RP,'ACQ:TRIG:DLY 8192');  % Delay
    writeline(RP,'ACQ:SOUR1:GAIN LV');  % Sets gain to LV/HV (should be the same as jumpers)
    writeline(RP,'ACQ:SOUR2:GAIN LV');  % Sets gain to LV/HV (should be the same as jumpers)
    
    % Start & Trigger
    % Trigger source setting must be after ACQ:START
    % Set trigger to source 1 positive edge

    writeline(RP,'ACQ:START');

    % After acquisition is started some time delay is needed in order to acquire fresh samples in the buffer
    pause(1);
    % Here we have used time delay of one second, but you can calculate the exact value by taking into account buffer
    % length and sampling rate

    writeline(RP,'ACQ:TRIG NOW');  % Instant data aquisition
        
    % Wait for trigger
    % Until trigger is true wait with acquiring
    % Be aware of the while loop if trigger is not achieved
    % Ctrl+C will stop code execution in MATLAB

    while 1
       trig_rsp = writeread(RP,'ACQ:TRIG:STAT?');
       if strcmp('TD',trig_rsp(1:2))  % Read only TD
           break;
       end
    end

    % Read data from buffer
    IN1 = writeread(RP,'ACQ:SOUR1:DATA?');
    IN2 = writeread(RP,'ACQ:SOUR2:DATA?');

    % Convert values to numbers.
    % The first character in string is "{" and the last 3 are 2 spaces and "}".
    
    IN1_num = str2num(IN1(1, 2:length(IN1)-3));
    DF_IN1(:,n) = IN1_num';

    IN2_num = str2num(IN2(1, 2:length(IN2)-3));
    DF_IN2(:,n) = IN2_num';
    
    % Turn off generator OUT1
    writeline(RP,'OUTPUT1:STATE OFF');
end

    
%% Close connection to Red Pitaya
clear RP;


%% Save data as mat file
save('./data/IN_INT.mat', 'DF_IN1', 'DF_IN2');
% save('./data/IN1_INT.mat', 'DF_IN1');
% save('./data/IN2_INT.mat', 'DF_IN2');

%% Save data as parquet file
% parquet data is of type table, no matrix operations
parquetwrite('data/IN1_INT.parquet', array2table(DF_IN1));
parquetwrite('data/IN2_INT.parquet', array2table(DF_IN2));

%% Save data as excel sheet
% data is table data, no matrix operations
% writematrix(DF_IN1, './data/IN1_UB_VBS_mat.xlsx');
% writematrix(DF_IN1, './data/IN2_UB_VBS_mat.xlsx');
% writematrix(DF_IN1, './data/IN_UB_VBS_VBP_mat.xlsx', 'Sheet', 1);
% writematrix(DF_IN2, './data/IN_UB_VBS_VBP_mat.xlsx', 'Sheet', 2);

File: 001_Analoge_Schaltungen/schaltungsentwurf_no1/redpitaya_scpi/bode_data_meas.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Measurements for Bode plots
Input signal: DF_IN1
Output signal: DF_IN2

@author: Mirco Meiners (HSB)
"""

# %% Init
import time
# from datetime import datetime
import redpitaya_scpi as scpi
import numpy as np
import pandas as pd
# import matplotlib.pyplot as plt

# %% Connection params
IP = '192.168.111.183'
rp_s = scpi.scpi(IP)

# %% Measurement / Data Accquisition
DF_IN1 = pd.DataFrame()
DF_IN2 = pd.DataFrame()

# Parameters
func = 'SINE'
ampl = 0.2
offset = 0.0
freqs = np.arange(10, 1000, 10)


for freq in freqs:

    rp_s.tx_txt('GEN:RST')  # Signal Generator reset
    rp_s.tx_txt('SOUR1:FUNC ' + str(func).upper())  # Wave form
    rp_s.tx_txt('SOUR1:VOLT ' + str(ampl))  # Magnitude
    rp_s.tx_txt('SOUR1:VOLT:OFFS ' + str(offset))  # Offset
    rp_s.tx_txt('SOUR1:FREQ:FIX ' + str(freq))  # Frequency
    rp_s.tx_txt('OUTPUT1:STATE ON')  # Output
    rp_s.tx_txt('SOUR1:TRig:INT')
    time.sleep(1)

    # Trigger
    rp_s.tx_txt('ACQ:RST')  # Input reset
    rp_s.tx_txt('ACQ:DEC 64')  # Decimation
    rp_s.tx_txt('ACQ:TRIG:LEV 0.5')  # Trigger level
    rp_s.tx_txt('ACQ:TRIG:DLY 8192')  # Delay
    rp_s.tx_txt('ACQ:START')  # Start measurement
    rp_s.tx_txt('ACQ:TRIG NOW')

    # Input IN1
    time.sleep(0.1)  # in seconds
    rp_s.tx_txt('ACQ:SOUR1:DATA?')  # Readout buffer IN1
    IN1str = rp_s.rx_txt()
    IN1str = IN1str.strip('{}\n\r').replace("  ", "").split(',')
    IN1 = np.array(list(map(float, IN1str)))
    DF_IN1[str(freq)] = IN1

    # Input IN2
    time.sleep(0.1)  # in seconds
    rp_s.tx_txt('ACQ:SOUR2:DATA?')  # Readout buffer IN2
    IN2str = rp_s.rx_txt()
    IN2str = IN2str.strip('{}\n\r').replace("  ", "").split(',')
    IN2 = np.array(list(map(float, IN2str)))
    DF_IN2[str(freq)] = IN2

    rp_s.tx_txt('OUTPUT2:STATE OFF')

# %% Data storage
Data_IN1 = 'data/IN1_INT_IN'  # + str(datetime.now().strftime('%Y-%m-%d_%H_%M'))
Data_IN2 = 'data/IN2_INT_OUT'  #+ str(datetime.now().strftime('%Y-%m-%d_%H_%M'))

# %% Store data on disk as comma-seperated-values
DF_IN1.to_csv(Data_IN1 + '.csv', index=False)
DF_IN2.to_csv(Data_IN2 + '.csv', index=False)

# %% Store data on disk as excel sheet
# with pd.ExcelWriter(Data_IN + '.xlsx') as writer:
#     DF_IN1.to_excel(writer, sheet_name='IN1', index=False)
#     DF_IN2.to_excel(writer, sheet_name='IN2', index=False)

# DF_IN1.to_excel(Data_IN1 + '.xlsx', index=False)
# DF_IN2.to_excel(Data_IN2 + '.xlsx', index=False)

# %% Store data on disk as HDF5
# DF_IN1.to_hdf(Data_IN1 + ".h5", "table", append=True)
# DF_IN2.to_hdf(Data_IN2 + ".h5", "table", append=True)

# %% Store data on disk as apache parquet
DF_IN1.to_parquet(Data_IN1 + ".parquet", index=False)
DF_IN2.to_parquet(Data_IN2 + ".parquet", index=False)

# %% Store data on disk as apache feather
# DF_IN1.to_feather(Data_IN1 + ".feather")
# DF_IN2.to_feather(Data_IN2 + ".feather")

# %% Plot
# plt.plot(DF_IN1['900'], label='IN1')
# plt.plot(DF_IN2['900'], label='IN2')
# plt.legend()
# plt.show()
